### PR TITLE
Refactor GIBBON to avoid retracing

### PR DESCRIPTION
**Pull request by uri-granta**

### CODE REVIEW

File: trieste/acquisition/function.py [hunk 1]
```diff
@@ -477,6 +464,8 @@ def __init__(self, model: ProbabilisticModel, samples: TensorType):
 
     def update(self, samples: TensorType) -> None:
         """Update the acquisition function with new samples."""
+        tf.debugging.assert_rank(samples, 2)
+        tf.debugging.assert_positive(len(samples))
         self._samples.assign(samples)
```
Comment (henrymoss) [inline trieste/acquisition/function.py:469]: I've added in a few of these extra asserts. Was there a reason not to have these before? I guess they will need unit testing now

File: trieste/acquisition/function.py [hunk 2]
```diff
@@ -1685,156 +1677,281 @@ def prepare_acquisition_function(
...
+        self,
+        model: ProbabilisticModel,
+        pending_points: TensorType,
+        rescaled_repulsion: bool = True,
+    ):
+        r"""
+        GIBBON's repulsion term encourages diversity within the batch
+        (achieving high values for points with low predictive correlation).
+
+        The term :math:`r=\log |C|` is given by the log determinant of the predictive
```
Comment (henrymoss) [inline trieste/acquisition/function.py]: the *replusion* term

File: trieste/acquisition/function.py [hunk 3]
```diff
@@ -1685,156 +1677,281 @@ def prepare_acquisition_function(
...
+            num_samples,
+            model,
+            dataset,
+            sample_min_value=True,
+            num_features=num_fourier_features,
+        )
+    else:  # use exact Thompson sampler
+        sampler = ExactThompsonSampler(num_samples, model, sample_min_value=True)
+
+    return sampler.sample(sampled_points)
```
Comment (henrymoss) [inline trieste/acquisition/function.py:1959]: This needs unit testing

Comment (uri-granta) [reply trieste/acquisition/function.py:1959]: These is a private function so wouldn't normally be unit tested directly. The code paths are tested instead via MinValueEntropySearch and GIBBON.

**Review by henrymoss — APPROVED** [review]
LGTM!

**Pull request merged**